_P172_PZEM-004Tv3.ino

```diff
@@ -0,0 +1,873 @@
+//#define USES_P172
+#ifdef USES_P172
+//#######################################################################################################
+//################### Plugin 172 PZEM-004Tv3 AC Current and Voltage measurement sensor ##################
+//#######################################################################################################
+/*
+    Needed modifications in ESPEasy source code
+        In _CPlugin_SensorTypeHelper.ino:
+        Insert at line 30 and 31
+                case SENSOR_TYPE_SEX:
+                    return 6;
+
+        In ESPEasy_checks.ino:
+        Change line 28:
+        check_size<ExtraTaskSettingsStruct,               472u>();
+        to:
+        check_size<ExtraTaskSettingsStruct,               640u>();
+
+        In DeviceStruct.h:
+        Insert at line 30:
+        #define SENSOR_TYPE_SEX                    66 
+
+        In Custom.h:
+        Add line:
+        #define VARS_PER_TASK 6
+        and:
+        #define USES_P172   // PZEM-004T version 3
+        And your own configuration settings ofcourse
+
+*/
+//#ifdef PLUGIN_BUILD_TESTING
+
+#include <ESPeasySerial.h>
+#include <vector>
+
+#define PLUGIN_172
+#define PLUGIN_ID_172           172
+
+#define PLUGIN_NAME_172         "Energy (AC) PZEM-004T version 3"
+#define PLUGIN_VALUENAME1_172   "Voltage"
+#define PLUGIN_VALUENAME2_172   "Current"
+#define PLUGIN_VALUENAME3_172   "Power"
+#define PLUGIN_VALUENAME4_172   "Energy"
+#define PLUGIN_VALUENAME5_172   "Frequency"
+#define PLUGIN_VALUENAME6_172   "PowerFactor"
+
+#define P172_REG_VOLTAGE        0x0000
+#define P172_REG_CURRENT_L      0x0001
+#define P172_REG_CURRENT_H      0X0002
+#define P172_REG_POWER_L        0x0003
+#define P172_REG_POWER_H        0x0004
+#define P172_REG_ENERGY_L       0x0005
+#define P172_REG_ENERGY_H       0x0006
+#define P172_REG_FREQUENCY      0x0007
+#define P172_REG_PF             0x0008
+#define P172_REG_ALARM          0x0009
+
+#define P172_CMD_RHR            0x03
+#define P172_CMD_RIR            0X04
+#define P172_CMD_WSR            0x06
+#define P172_CMD_CAL            0x41
+#define P172_CMD_REST           0x42
+
+
+#define P172_WREG_ALARM_THR     0x0001
+#define P172_WREG_ADDR          0x0002
+
+#define P172_UPDATE_TIME        200
+
+#define P172_RESPONSE_SIZE      32
+#define P172_READ_TIMEOUT       200
+#define P172_DELAY_TIMEOUT      2000
+#define P172_ERROR_VALUE        -1
+#define P172_PZEM_BAUD_RATE     9600
+
+
+ESPeasySerial *P172_easySerial = nullptr;
+static bool P172_first_init_done = false;
+static int P172_rx_pin = 0;
+static int P172_tx_pin = 0;
+
+static uint8_t P172_first_unit_index = 255;
+uint64_t P172_lastRead; 
+unsigned long P172_timeout_msec;
+
+struct P172_PZEM004T_values
+{
+    taskIndex_t task_index;
+    float voltage;
+    float current;
+    float power;
+    float energy;
+    float frequency;
+    float pf;
+    uint16_t alarms;
+    uint8_t slave_addr_check;
+    uint8_t slave_addr;
+};  // Measured values

```

Comment:
It is good practice to sort members of a struct based on the size they occupy, with the largest ones first.
This will reduce the total amount of memory used.
Especially when you make a vector or an array of those structs, then every byte you gain may matter.